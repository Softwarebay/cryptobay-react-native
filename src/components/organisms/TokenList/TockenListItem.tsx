import React from 'react';
import { View } from 'react-native';

import { Token } from '../../../types/token.type';
import { Badge, Caption, Label, Picture } from '../../atoms';

import { tockenListItemStyles } from './TokenList.styles';

export const TokenListItem: React.FC<Token> = ({
  picture,
  name,
  shortName,
  price,
  percentagePoint,
  value,
}: Token) => {
  const captionLabel = `$${price}`;
  const percentageLabel =
    percentagePoint > 0 ? `+${percentagePoint}%` : `${percentagePoint}%`;
  const variant = percentagePoint > 0 ? 'green' : 'red';
  const valueLabel = `${value} ${shortName}`;

  return (
    <View style={tockenListItemStyles.container}>
      <View style={tockenListItemStyles.dataWrapper}>
        <Picture style={tockenListItemStyles.picture} pictureFile={picture} />

        <View>
          <Label value={name} />

          <View style={tockenListItemStyles.valueData}>
            <Caption
              style={tockenListItemStyles.caption}
              value={captionLabel}
            />
            <Badge value={percentageLabel} variant={variant} />
          </View>
        </View>
      </View>

      <Label value={valueLabel} />
    </View>
  );
};
