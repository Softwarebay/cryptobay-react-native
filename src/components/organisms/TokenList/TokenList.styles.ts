import { StyleSheet } from 'react-native';

export const tockenListItemStyles = StyleSheet.create({
  container: {
    marginVertical: 16,
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  dataWrapper: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  picture: {
    marginRight: 8,
    width: 40,
    height: 40,
    borderRadius: 40,
    backgroundColor: '#44485F',
  },
  valueData: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  caption: {
    marginRight: 8,
  },
});

export const tockenListStyles = StyleSheet.create({
  container: {
    width: '100%',
  },
  nav: {
    width: '50%',
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  navItem: {
    marginHorizontal: 10,
    paddingHorizontal: 15,
  },
  previousTab: {
    color: '#888DAA',
  },
  currentTab: {
    borderBottomWidth: 2,
    borderColor: '#FFFFFF',
  },
  list: {
    minHeight: 350,
  },
});
