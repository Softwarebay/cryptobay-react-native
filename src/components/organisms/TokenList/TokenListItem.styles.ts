import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
  container: {
    marginVertical: 16,
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  dataWrapper: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  picture: {
    marginRight: 8,
    width: 40,
    height: 40,
    borderRadius: 40,
    backgroundColor: '#44485F',
  },
  valueData: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  caption: {
    marginRight: 8,
  },
});
