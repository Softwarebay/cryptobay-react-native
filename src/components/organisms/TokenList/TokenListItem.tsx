import { useNavigation } from '@react-navigation/native';
import React from 'react';
import { TouchableOpacity, View } from 'react-native';

import { Screens } from '../../../screens';
import { Token } from '../../../types/token.type';
import { Badge, Caption, Label, Picture } from '../../atoms';

import { styles } from './TokenListItem.styles';

export const TokenListItem: React.FC<Token> = ({
  picture,
  name,
  shortName,
  price,
  percentagePoint,
  value,
}: Token) => {
  const navigation = useNavigation();

  const tokenPicture = picture ? picture : 0;
  const captionLabel = price % 1 === 0 ? `$${price}.00` : `$${price}`;
  const percentageLabel =
    percentagePoint > 0 ? `+${percentagePoint}%` : `${percentagePoint}%`;
  const variant = percentagePoint > 0 ? 'green' : 'red';
  const valueLabel = `${value} ${shortName}`;
  const token = {
    shortName: shortName,
    value: value,
    price: price,
  };

  const onTokenNavigate = () =>
    navigation.navigate(Screens.TOKEN as never, { token: token } as never);

  return (
    <TouchableOpacity style={styles.container} onPress={onTokenNavigate}>
      <View style={styles.dataWrapper}>
        <Picture style={styles.picture} pictureFile={tokenPicture} />

        <View>
          <Label value={name} />

          <View style={styles.valueData}>
            <Caption style={styles.caption} value={captionLabel} />
            <Badge value={percentageLabel} variant={variant} />
          </View>
        </View>
      </View>

      <Label value={valueLabel} />
    </TouchableOpacity>
  );
};
