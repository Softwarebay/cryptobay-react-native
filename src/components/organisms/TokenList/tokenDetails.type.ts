export type TokenDetails = {
  shortName: string;
  value: number;
  price: number;
};
