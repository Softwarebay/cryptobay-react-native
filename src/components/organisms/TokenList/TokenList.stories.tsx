import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react-native';
import React from 'react';

import StorybookView from '../../../../storybook/StorybookView';
import { tokensMock } from '../../../mocks';

import { TokenList } from './TokenList';

storiesOf('Organisms/TokenList', module)
  .addDecorator(getStory => <StorybookView>{getStory()}</StorybookView>)
  .add('Token List', () => (
    <TokenList
      tokens={tokensMock}
      onTokenNavigate={action('onTokenNavigate')}
    />
  ));
