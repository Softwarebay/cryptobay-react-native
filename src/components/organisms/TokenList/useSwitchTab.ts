import React from 'react';

import { styles } from './TokenList.styles';

enum Tabs {
  TOKEN = 'TOKEN',
  COLLECTIBILES = 'COLLECTIBILES',
}

export const useSwitchTab = () => {
  const [currentTab, setCurrentTab] = React.useState<Tabs>(Tabs.TOKEN);

  const handleSwitchTab = (tab: Tabs) => setCurrentTab(tab);

  const handleSwitchTabStyle = (tab: Tabs) =>
    tab === currentTab && styles.currentTab;

  const handleSwitchLabelStyles = (tab: Tabs) =>
    tab !== currentTab && styles.previousTab;

  return {
    Tabs,
    currentTab,
    handleSwitchTab,
    handleSwitchTabStyle,
    handleSwitchLabelStyles,
  };
};
