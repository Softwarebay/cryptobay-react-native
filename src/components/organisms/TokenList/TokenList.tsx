import React from 'react';
import {
  FlatList,
  ListRenderItemInfo,
  TouchableOpacity,
  View,
} from 'react-native';

import { Token } from '../../../types/token.type';
import { Label } from '../../atoms';

import { TokenListItem } from './TockenListItem';
import { tockenListStyles } from './TokenList.styles';
import { useSwitchTab } from './useSwitchTab';

type TokenListProps = {
  tokens: Token[];
};

export const TokenList: React.FC<TokenListProps> = ({
  tokens,
}: TokenListProps) => {
  const {
    Tabs,
    currentTab,
    handleSwitchTab,
    handleSwitchTabStyle,
    handleSwitchLabelStyles,
  } = useSwitchTab();

  const renderCurrentTab = {
    TOKEN: (
      <FlatList
        style={tockenListStyles.list}
        data={tokens}
        renderItem={({ item }: ListRenderItemInfo<Token>) => (
          <TokenListItem
            picture={item.picture}
            name={item.name}
            shortName={item.shortName}
            price={item.price}
            percentagePoint={item.percentagePoint}
            value={item.value}
          />
        )}
      />
    ),
    COLLECTIBILES: (
      <FlatList
        style={tockenListStyles.list}
        data={['Collectibiles in Progress']}
        renderItem={({ item }: ListRenderItemInfo<string>) => (
          <Label value={item} />
        )}
      />
    ),
  };

  return (
    <View style={tockenListStyles.container}>
      <View style={tockenListStyles.nav}>
        <TouchableOpacity
          style={[tockenListStyles.navItem, handleSwitchTabStyle(Tabs.TOKEN)]}
          onPress={() => handleSwitchTab(Tabs.TOKEN)}>
          <Label style={handleSwitchLabelStyles(Tabs.TOKEN)} value="Token" />
        </TouchableOpacity>

        <TouchableOpacity
          style={[
            tockenListStyles.navItem,
            handleSwitchTabStyle(Tabs.COLLECTIBILES),
          ]}
          onPress={() => handleSwitchTab(Tabs.COLLECTIBILES)}>
          <Label
            style={handleSwitchLabelStyles(Tabs.COLLECTIBILES)}
            value="Collectibles"
          />
        </TouchableOpacity>
      </View>

      {renderCurrentTab[currentTab]}
    </View>
  );
};
