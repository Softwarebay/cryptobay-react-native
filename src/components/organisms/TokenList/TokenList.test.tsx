import { fireEvent, render, screen } from '@testing-library/react-native';
import React from 'react';
import '@testing-library/jest-native';
import renderer from 'react-test-renderer';

import 'react-native';
import { PROFILE } from '../../../../assets/pictures';
import { tokensMock } from '../../../mocks';

import { TokenList } from './TokenList';
import { TokenListItem } from './TokenListItem';

const mockOnPressToken = jest.fn();

describe('TokenList Organism', () => {
  test('Component Snapshot', () => {
    const tree = renderer
      .create(<TokenList tokens={tokensMock} onPressToken={mockOnPressToken} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('ensure changing tabs works correctly', () => {
    render(<TokenList tokens={tokensMock} onPressToken={mockOnPressToken} />);
    const tokenTab = screen.getByTestId('token-list-token-tab');
    const collectiblesTab = screen.getByTestId('token-list-collectibles-tab');
    const currentTabstyle = {
      borderBottomWidth: 2,
      borderColor: '#FFFFFF',
    };
    fireEvent.press(tokenTab);
    expect(tokenTab).toHaveStyle(currentTabstyle);
    fireEvent.press(collectiblesTab);
    expect(collectiblesTab).toHaveStyle(currentTabstyle);
  });

  test('ensure token list navigate works', () => {
    render(
      <TokenListItem
        name="Binance Coin"
        shortName="BNB"
        price={226.69}
        percentagePoint={226.69}
        value={19.2371}
        testID="token-list-item"
        onPressToken={mockOnPressToken}
      />,
    );
    fireEvent.press(screen.getByTestId('token-list-item'));
    expect(mockOnPressToken).toHaveBeenCalledWith({
      shortName: 'BNB',
      value: 19.2371,
      price: 226.69,
    });
  });

  test('ensure token list item picture works correctly', () => {
    const { rerender } = render(
      <TokenListItem
        name="Binance Coin"
        shortName="BNB"
        price={226.69}
        percentagePoint={226.69}
        value={19.2371}
        testID="token-list-item"
        onPressToken={mockOnPressToken}
      />,
    );
    const tokenPicture = screen.getByTestId('token-list-item-token-picture');
    expect(tokenPicture).toHaveProp('source', 0);
    rerender(
      <TokenListItem
        picture={PROFILE}
        name="Binance Coin"
        shortName="BNB"
        price={226.69}
        percentagePoint={226.69}
        value={19.2371}
        testID="token-list-item"
        onPressToken={mockOnPressToken}
      />,
    );
    expect(tokenPicture).toHaveProp('source', PROFILE);
  });
});
