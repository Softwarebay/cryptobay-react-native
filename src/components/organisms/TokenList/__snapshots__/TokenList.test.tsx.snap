// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenList Organism Component Snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": "50%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "borderBottomWidth": 2,
          "borderColor": "#FFFFFF",
          "marginHorizontal": 10,
          "opacity": 1,
          "paddingHorizontal": 15,
        }
      }
      testID="token-list-token-tab"
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Poppins",
            },
            Object {
              "fontSize": 16,
              "fontWeight": "600",
              "lineHeight": 24,
            },
            Object {
              "color": "#FFFFFF",
            },
          ]
        }
      >
        Token
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginHorizontal": 10,
          "opacity": 1,
          "paddingHorizontal": 15,
        }
      }
      testID="token-list-collectibles-tab"
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Poppins",
            },
            Object {
              "fontSize": 16,
              "fontWeight": "600",
              "lineHeight": 24,
            },
            Object {
              "color": "#888DAA",
            },
          ]
        }
      >
        Collectibles
      </Text>
    </View>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "name": "Binance Coin",
          "percentagePoint": 2,
          "picture": 0,
          "price": 226.69,
          "shortName": "BNB",
          "value": 19.2371,
        },
        Object {
          "name": "USD Coin",
          "percentagePoint": 4.3,
          "picture": 0,
          "price": 1,
          "shortName": "USDC",
          "value": 92.3,
        },
        Object {
          "name": "Synthetix",
          "percentagePoint": -1.3,
          "picture": 0,
          "price": 20.83,
          "shortName": "SNX",
          "value": 42.74,
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "height": 250,
      }
    }
    testID="token-list"
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginVertical": 16,
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Image
              source={0}
              style={
                Object {
                  "backgroundColor": "#44485F",
                  "borderRadius": 40,
                  "height": 40,
                  "marginRight": 8,
                  "width": 40,
                }
              }
              testID="token-list-item-token-picture"
            />
            <View>
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": "Poppins",
                    },
                    Object {
                      "fontSize": 16,
                      "fontWeight": "600",
                      "lineHeight": 24,
                    },
                    Object {
                      "color": "#FFFFFF",
                    },
                  ]
                }
              >
                Binance Coin
              </Text>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "Poppins",
                        "fontSize": 12,
                        "fontWeight": "400",
                        "lineHeight": 18,
                      },
                      Object {
                        "color": "#ABAFC4",
                      },
                      Object {
                        "marginRight": 8,
                      },
                    ]
                  }
                >
                  $226.69
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "Poppins",
                        "fontSize": 14,
                        "fontWeight": "600",
                        "lineHeight": 24,
                      },
                      Object {
                        "color": "#76E268",
                      },
                      undefined,
                    ]
                  }
                >
                  +2%
                </Text>
              </View>
            </View>
          </View>
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Poppins",
                },
                Object {
                  "fontSize": 16,
                  "fontWeight": "600",
                  "lineHeight": 24,
                },
                Object {
                  "color": "#FFFFFF",
                },
              ]
            }
          >
            19.2371 BNB
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginVertical": 16,
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Image
              source={0}
              style={
                Object {
                  "backgroundColor": "#44485F",
                  "borderRadius": 40,
                  "height": 40,
                  "marginRight": 8,
                  "width": 40,
                }
              }
              testID="token-list-item-token-picture"
            />
            <View>
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": "Poppins",
                    },
                    Object {
                      "fontSize": 16,
                      "fontWeight": "600",
                      "lineHeight": 24,
                    },
                    Object {
                      "color": "#FFFFFF",
                    },
                  ]
                }
              >
                USD Coin
              </Text>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "Poppins",
                        "fontSize": 12,
                        "fontWeight": "400",
                        "lineHeight": 18,
                      },
                      Object {
                        "color": "#ABAFC4",
                      },
                      Object {
                        "marginRight": 8,
                      },
                    ]
                  }
                >
                  $1.00
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "Poppins",
                        "fontSize": 14,
                        "fontWeight": "600",
                        "lineHeight": 24,
                      },
                      Object {
                        "color": "#76E268",
                      },
                      undefined,
                    ]
                  }
                >
                  +4.3%
                </Text>
              </View>
            </View>
          </View>
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Poppins",
                },
                Object {
                  "fontSize": 16,
                  "fontWeight": "600",
                  "lineHeight": 24,
                },
                Object {
                  "color": "#FFFFFF",
                },
              ]
            }
          >
            92.3 USDC
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginVertical": 16,
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Image
              source={0}
              style={
                Object {
                  "backgroundColor": "#44485F",
                  "borderRadius": 40,
                  "height": 40,
                  "marginRight": 8,
                  "width": 40,
                }
              }
              testID="token-list-item-token-picture"
            />
            <View>
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": "Poppins",
                    },
                    Object {
                      "fontSize": 16,
                      "fontWeight": "600",
                      "lineHeight": 24,
                    },
                    Object {
                      "color": "#FFFFFF",
                    },
                  ]
                }
              >
                Synthetix
              </Text>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "Poppins",
                        "fontSize": 12,
                        "fontWeight": "400",
                        "lineHeight": 18,
                      },
                      Object {
                        "color": "#ABAFC4",
                      },
                      Object {
                        "marginRight": 8,
                      },
                    ]
                  }
                >
                  $20.83
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "Poppins",
                        "fontSize": 14,
                        "fontWeight": "600",
                        "lineHeight": 24,
                      },
                      Object {
                        "color": "#EA3943",
                      },
                      undefined,
                    ]
                  }
                >
                  -1.3%
                </Text>
              </View>
            </View>
          </View>
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Poppins",
                },
                Object {
                  "fontSize": 16,
                  "fontWeight": "600",
                  "lineHeight": 24,
                },
                Object {
                  "color": "#FFFFFF",
                },
              ]
            }
          >
            42.74 SNX
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
